File: D/979-distribute-coins-in-binary-tree/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func distributeCoins(root *TreeNode) int {
	ret := 0
	var recur func(root *TreeNode) int
	recur = func(root *TreeNode) int {
		if root.Left == nil && root.Right == nil {
			d := root.Val - 1
			if d < 0 {
				d = -d
			}
			ret += d
			return root.Val - 1
		}

		left, right := 0, 0
		if root.Left != nil {
			left = recur(root.Left)
		}
		if root.Right != nil {
			right = recur(root.Right)
		}
		d := left + right + root.Val - 1
		if d < 0 {
			d = -d
		}
		ret += d
		return left + right + root.Val - 1
	}
	recur(root)
	return ret
}

func main() {
	fmt.Println()
}

File: D/979-distribute-coins-in-binary-tree/main.rs
struct Solution {}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
  pub fn distribute_coins(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    fn recur(root: Option<Rc<RefCell<TreeNode>>>, cnt: &mut i32) -> i32 {
      if let Some(node) = root {
        let left = recur(node.borrow_mut().left.take(), cnt);
        let right = recur(node.borrow_mut().right.take(), cnt);
        let v = left + right + node.borrow().val - 1;
        *cnt += v.abs();
        v
      } else {
        0
      }
    }
    let mut cnt: i32 = 0;
    recur(root, &mut cnt);
    cnt
  }
}
